This folder shows examples of running the whole workflow and running only the visualization.

#### Workflow
Folder example/workflow includes input examples for running the whole workflow, where:

- Reference: the folder stores reference genome.
- Annotations: the folder stores annotation file. Annotation file is in csv format and should have the format of seven columns: 
	- Name: gene name
	- Type: can be rRNA, repreat_region, coding sequence, tRNA,... This annotation file is used in visualization.
	- Minimum: the start position of the gene on reference genome
	- Maximum: the end position of the gene on reference genome
	- Length: length of gene, literally, length = end - start + 1
	- \# Intervals: number of intervals of the gene.
	- Direction: forwad or reward.
- config.txt: shows the example of config file for the workflow.
- readids.txt: show the example of how to specify read IDs. If there are fastq files named SRR2146923_1.fastq, SRR2146923_2.fastq, SRR2146925_1.fastq, SRR2146925_2.fastq, only the IDs SRR2146923 and SRR2146925 are needed in the readids file.



#### Visualization:
Folder example/visualization includes examples for running the visualization phase.

The inputs for this phase include:
- Genome name.
- Annotation file.
- Heteroplasmy file: in csv format. This file is generated by the workflow.
- Conserved site file: in csv format. This file is generated by the workflow.
- Output filename: in html format. 
